src/adapters/nats.js

```diff
@@ -0,0 +1,436 @@
+	 * @param {StreamConfig} streamOpts JetStream stream configs
+	 */
+	async createStream(streamName, subjects, streamOpts) {
+		const streamConfig = _.defaultsDeep(
+			{
+				name: streamName,
+				subjects: subjects
+			},
+			streamOpts,
+			this.opts.nats.streamConfig
+		);
+
+		try {
+			const streamInfo = await this.manager.streams.add(streamConfig);
+			this.logger.debug(streamInfo);
+		} catch (error) {
+			if (error.message === "stream name already in use") {
+				// Silently ignore the error. Channel or Consumer Group already exists
+				this.logger.debug(`NATS Stream with name: '${streamName}' already exists.`);
+			} else {
+				this.logger.error("An error ocurred while create NATS Stream", error);
+			}
+		}
+	}
+
+	/**
+	 * Moves message into dead letter
+	 *
+	 * @param {Channel} chan
+	 * @param {JsMsg} message JetStream message
+	 */
+	async moveToDeadLetter(chan, message) {
+		// this.logger.warn(`Moved message to '${chan.deadLettering.queueName}'`);
+
+		try {
+			await this.publish(chan.deadLettering.queueName, message.data, { raw: true });
+
+			this.logger.warn(`Moved message to '${chan.deadLettering.queueName}'`, message.seq);
+		} catch (error) {
+			this.logger.info("An error occurred while moving", error);
+		}
+	}
+
+	/**
+	 * Unsubscribe from a channel.
+	 *
+	 * @param {Channel} chan
+	 */
+	async unsubscribe(chan) {
+		chan.unsubscribing = true;
+		const sub = this.subscriptions.get(chan.id);
+		if (!sub) return;
+
+		await new Promise((resolve, reject) => {
+			const checkPendingMessages = () => {
+				try {
+					if (this.getNumberOfChannelActiveMessages(chan.id) === 0) {
+						// More info: https://github.com/nats-io/nats.deno/blob/main/jetstream.md#push-subscriptions
+						return sub
+							.drain()
+							.then(() => sub.unsubscribe())
+							.then(() => {
+								this.logger.debug(
+									`Unsubscribing from '${chan.name}' chan with '${chan.group}' group...'`
+								);
+
+								// Stop tracking channel's active messages
+								this.stopChannelActiveMessages(chan.id);
+
+								resolve();
+							})
+							.catch(err => reject(err));
+					} else {
+						this.logger.warn(
+							`Processing ${this.getNumberOfChannelActiveMessages(
+								chan.id
+							)} message(s) of '${chan.id}'...`
+						);
+
+						setTimeout(() => checkPendingMessages(), 1000);
+					}
+				} catch (err) {
+					reject(err);
+				}
+			};
+
+			checkPendingMessages();
+		});
+	}
+
+	/**
+	 * Publish a payload to a channel.
+	 *
+	 * @param {String} channelName
+	 * @param {any} payload
+	 * @param {Partial<JetStreamPublishOptions>?} opts
+	 */
+	async publish(channelName, payload, opts = {}) {
+		// Adapter is stopping. Publishing no longer is allowed
+		if (this.stopping) return;
+
+		// this.logger.info(`NATS out => topic:${channelName} || messageID:${payload.id}`);
+
+		try {
+			const response = await this.client.publish(
+				channelName,
+				opts.raw ? payload : this.serializer.serialize(payload),
+				opts
```

Comment:
The `headers` should be similar for all adapters. As I see JetStream has a custom format, so we should remap from object.